# Revision - Level 2 - 05/05/2026
Reinforce:
- what an API is
- requests (GET / POST)
- status codes
- response.json()
- basic API thinking

In [31]:
import requests

## ⚔️ EXERCISE 1 — Basic GET Request

👉 Task:
- Use requests to call:
  https://jsonplaceholder.typicode.com/posts/1

👉 Tasks:
- Print the status code
- Print the JSON response

👉 Question:
- What type is the response after using .json()?

In [32]:
url = "https://jsonplaceholder.typicode.com/posts/1"

response = requests.get(url)

print(response.status_code)
print(type(response))
print(response.json())
print(type(response.json()))

200
<class 'requests.models.Response'>
{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}
<class 'dict'>


The response become from a requests.model.Response to a dict

## ⚔️ EXERCISE 2 — Handling Status Code

👉 Task:
- Make the same GET request

👉 Add:
- If status_code == 200 → print "Success"
- Else → print "Error"

👉 Goal:
- Understand basic response handling

In [33]:
url = "https://jsonplaceholder.typicode.com/posts/1"

response = requests.get(url)

if response.status_code == 200:
    print("Success")
else:
    print("Error")

Success


## ⚔️ EXERCISE 3 — Extract Data from JSON

Using the same API response:

👉 Task:
- Extract and print:
  - title
  - body

👉 Goal:
- Navigate JSON like a dictionary

In [34]:
try:
    data = response.json()
    title = data['title']
    body = data['body']
except Exception as err:
    print(f"An error occurred {err}")

print("Title: ", title)
print("Body:", body)

Title:  sunt aut facere repellat provident occaecati excepturi optio reprehenderit
Body: quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto


## ⚔️ EXERCISE 4 — POST Request

👉 Task:
- Send a POST request to:
  https://jsonplaceholder.typicode.com/posts

👉 Data:
{
  "title": "test",
  "body": "learning API",
  "userId": 1
}

👉 Tasks:
- Print status code
- Print response JSON

👉 Question:
- What status code did you get? Why?

In [35]:
url = "https://jsonplaceholder.typicode.com/posts"
data = { "title": "test", "body": "learning API", "userId": 1 }

response = requests.post(url, json=data)

print(response.status_code)

data = response.json()
print(data)

201
{'title': 'test', 'body': 'learning API', 'userId': 1, 'id': 101}


I got a status code of 201, which indicates that the request succeeded and created a new resource.

## ⚔️ EXERCISE 5 — Concept Check

👉 Answer:

1. What is an API?
2. Difference between:
   - GET
   - POST
3. What does response.json() do?
4. What does status code 200 mean?
5. Why is error handling important?

- API stands for **Application Program Interface** and it's a program that handles communication between backend and frontend, using HTTPS requests
- GET it's used to retrieve information, POST is used to create information
- `response.json()` returns data if the request is successful, which means that the request was able to retrieve information from an specific site
- Status code 200 means the request was succssesful
- Because bad requests happens a lot, so you can have multiple errors when dealing with APIs, such as internet problems, non existent requests, bad requests, which are requests that are malformed syntax or is inherently incorrect in its format.

## ⚔️ TASK

👉 Create a function:

fetch_post()

---

## ⚔️ REQUIREMENTS

Inside the function:

1. Make a GET request to:
   https://jsonplaceholder.typicode.com/posts

2. Check if status_code == 200

3. Convert response to JSON

4. Print:
   - title
   - body

In [45]:
def get_posts(number_of_posts):
    url = "https://jsonplaceholder.typicode.com/posts"
    response = requests.get(url)
    posts_list = []
    try:
        posts = response.json()
        for post in posts[:number_of_posts]:
            posts_list.append(post)
        return posts_list
    except Exception as err:
        return f"An error occurred {err}"

posts = get_posts(3)
for post in posts:
    print(f"ID: {post['id']}\nTitle: {post['title']}\nBody: {post['body']}\n")

ID: 1
Title: sunt aut facere repellat provident occaecati excepturi optio reprehenderit
Body: quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto

ID: 2
Title: qui est esse
Body: est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla

ID: 3
Title: ea molestias quasi exercitationem repellat qui ipsa sit aut
Body: et iusto sed quo iure
voluptatem occaecati omnis eligendi aut ad
voluptatem doloribus vel accusantium quis pariatur
molestiae porro eius odio et labore et velit aut

